libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/building/BuildingProcessor.kt

```diff
@@ -0,0 +1,58 @@
+package com.mapbox.navigation.ui.maps.building
+
+import com.mapbox.bindgen.ExpectedFactory
+import com.mapbox.maps.RenderedQueryOptions
+import com.mapbox.navigation.ui.maps.building.model.BuildingError
+import com.mapbox.navigation.ui.maps.building.model.BuildingValue
+import kotlin.coroutines.resume
+import kotlin.coroutines.suspendCoroutine
+
+internal object BuildingProcessor {
+
+    /**
+     * Source layer that includes building height.
+     */
+    private const val BUILDING_LAYER_ID = "building"
+
+    suspend fun queryBuilding(
+        action: BuildingAction.QueryBuilding
+    ): BuildingResult.QueriedBuildings = suspendCoroutine { continuation ->
+        val screenCoordinate = action.mapboxMap.pixelForCoordinate(action.point)
+        val queryOptions = RenderedQueryOptions(listOf(BUILDING_LAYER_ID), null)
+        action.mapboxMap.queryRenderedFeatures(screenCoordinate, queryOptions) { expected ->
+            expected.fold(
+                { error ->
+                    continuation.resume(
+                        BuildingResult.QueriedBuildings(
+                            ExpectedFactory.createError(BuildingError(error, null))
+                        )
+                    )
+                },
+                { value ->
+                    continuation.resume(
+                        BuildingResult.QueriedBuildings(
+                            ExpectedFactory.createValue(BuildingValue(value))
+                        )
+                    )
+                }
+            )
+        }
+    }
+
+    fun queryBuildingOnWaypoint(
+        action: BuildingAction.QueryBuildingOnWaypoint
+    ): BuildingResult.GetDestination {
+        val routeOptions = action.progress.route.routeOptions()
+        val coordinateIndex = action.progress.currentLegProgress?.legIndex!! + 1
+        return routeOptions?.coordinatesList()?.get(coordinateIndex)?.let { point ->
```

Comment:
We should first check if there's a defined `"waypoint_targets"` and prioritize that.

In the best possible scenario, the integrating code would first use search capabilities to find a routing point and use that in the request. If that would've been the case, this code would never return anything. That's why, as mentioned previously, we should recommend setting `"waypoint_targets"` in the Directions request when using the building highlighting feature, and we should prioritize this field over `coordinatesList` when possible.

Reply:
This is not a blocker for this PR though. If not addressed here, let's cut a new ticket.

Reply:
To avoid potential out of bounds exceptions, should we use `getOrNull`?

```suggestion
        return routeOptions?.coordinatesList()?.getOrNull(coordinateIndex)?.let { point ->
```

Reply:
As per internal discussion with @LukasPaczos this would be addressed in a separate PR.